addons/barcodes/static/src/js/barcode_events.js

```
@@ -0,0 +1,148 @@
+        this.__handler = _.bind(this.handler, this);
+        // Bind event handler once the DOM is loaded
+        $(this.start.bind(this));
+    },
+
+    handle_buffered_keys: function() {
+        var str = _.reduce(this.buffered_key_events, function(memo, e) { return memo + String.fromCharCode(e.which) }, '');
+        var match = str.match(this.regexp);
+
+        if (match) {
+            var barcode = match[1];
+
+            core.bus.trigger('barcode_scanned', barcode);
+
+            // Dispatch a barcode_scanned DOM event to elements that have barcode_events="true" set.
+            if (this.buffered_key_events[0].target.getAttribute("barcode_events") === "true")
+                $(this.buffered_key_events[0].target).trigger('barcode_scanned', barcode);
+        } else {
+            this.resend_buffered_keys();
+        }
+
+        this.buffered_key_events = [];
+    },
+
+    resend_buffered_keys: function() {
+        function send_keys(list) {
+            if (list.length === 0) return;
+            var old_event = list.shift();
+            var $element = $(old_event.target);
+            var keyCode = old_event.keyCode || old_event.which;
+            setTimeout(function () {
+                setTimeout(function () {
+                    $element.trigger({ type: 'keydown', keyCode: keyCode });
+                    if ((keyCode > 47 && keyCode < 58)   || // number keys
+                        keyCode == 32  || // spacebar
+                        (keyCode > 64 && keyCode < 91)   || // letter keys
+                        (keyCode > 95 && keyCode < 112)  || // numpad keys
+                        (keyCode > 185 && keyCode < 193) || // ;=,-./` (in order)
+                        (keyCode > 218 && keyCode < 223)) {   // [\]' (in order))
+                        document.execCommand('insertText', 0, String.fromCharCode(keyCode));
+                    }
+                    $element.trigger({ type: 'keyup', keyCode: keyCode });
```

Why manually trigger keyup/down. What problems does that solve? As far as I can tell the native event handlers of INPUT, ... on Firefox and Chromium will just silently ignore these anyway. [Fiddle to test](http://jsfiddle.net/banana_apple/x3jeL3dq/3/).